sklearn/cluster/_hdbscan/_boruvka.pyx

```diff
@@ -0,0 +1,769 @@
+# up with sub-quadratic complexity at worst.
+#
+# This code is based on the papers:
+#
+# Fast Euclidean Minimum Spanning Tree: Algorithm, analysis, and applications
+# William B. March, Parikshit Ram, Alexander Gray
+# Conference: Proceedings of the 16th ACM SIGKDD International Conference on
+#  Knowledge Discovery and Data Mining
+# 2010
+#
+# Tree-Independent Dual-Tree Algorithms
+# Ryan R. Curtin, William B. March, Parikshit Ram, David V. Anderson,
+# Alexander G. Gray, Charles L. Isbell Jr
+# 2013, arXiv 1304.4327
+#
+# As per the sklearn BallTree and KDTree implementations we make use of
+# the rdist for KDTree, which is a faster-to-compute notion of distance
+# (for example in the euclidean case it is the distance squared).
+#
+# To combine together components in between sweeps we make use of
+# a union find data structure. This is a separate implementation
+# from that used in the labelling of the single linkage tree as
+# we can perform more specific optimizations here for what
+# is a simpler version of the structure.
+
+import numpy as np
+
+cimport numpy as cnp
+from libc.float cimport DBL_MAX
+from libc.math cimport fabs, pow
+
+from sklearn.neighbors import KDTree
+
+from ...metrics._dist_metrics cimport DistanceMetric, DistanceMetric64
+from ...utils._typedefs cimport intp_t, float64_t, uint8_t, int8_t
+from ...neighbors._binary_tree cimport NodeData_t
+from ._linkage cimport MST_edge_t
+from ._linkage import MST_edge_dtype
+from joblib import effective_n_jobs
+
+cdef float64_t INF = np.inf
+
+# Define a function giving the minimum distance between two
+# nodes of a ball tree
+cdef inline float64_t ball_tree_min_dist_dual(
+    float64_t radius1,
+    float64_t radius2,
+    intp_t node1,
+    intp_t node2,
+    float64_t[:, ::1] centroid_dist
+) noexcept nogil:
+
+    cdef float64_t dist_pt = centroid_dist[node1, node2]
+    return max(0, (dist_pt - radius1 - radius2))
+
+
+# Define a function giving the minimum distance between two
+# nodes of a kd-tree
+cdef inline float64_t kd_tree_min_dist_dual(
+    DistanceMetric64 metric,
+    intp_t node1,
+    intp_t node2,
+    float64_t[:, :, ::1] node_bounds,
+    intp_t num_features
+) noexcept nogil:
+
+    cdef float64_t d, d1, d2, rdist = 0.0
+    cdef intp_t j
+
+    if metric.p == INF:
+        for j in range(num_features):
+            d1 = node_bounds[0, node1, j] - node_bounds[1, node2, j]
+            d2 = node_bounds[0, node2, j] - node_bounds[1, node1, j]
+            d = (d1 + fabs(d1)) + (d2 + fabs(d2))
+
+            rdist = max(rdist, 0.5 * d)
+    else:
+        # here we'll use the fact that x + abs(x) = 2 * max(x, 0)
+        for j in range(num_features):
+            d1 = node_bounds[0, node1, j] - node_bounds[1, node2, j]
+            d2 = node_bounds[0, node2, j] - node_bounds[1, node1, j]
+            d = (d1 + fabs(d1)) + (d2 + fabs(d2))
+
+            rdist += pow(0.5 * d, metric.p)
+
+    return metric._rdist_to_dist(rdist)
+
+
+cdef class BoruvkaUnionFind:
```

Comment:
Note for reviewers: I wasn't sure of a "nice" way to reuse the `UnionFind` structure from `_hierarchical_fast` since this one explicitly keeps track of initial nodes as components without leveraging virtual nodes created through union operations, and thus the `UnionFind` structure would need to be modified with non-trivial additions. I think it is better to leave them as two similar-yet-different structures. Open to suggestions though!